src/backend/optimizer/path/allpaths.c

```diff
@@ -2012,7 +2012,12 @@ set_cte_pathlist(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
 
 		config->honor_order_by = false;
 
-		if (!cte->cterecursive)
+		/*
+		 * Additionally to recursive CTE queries, don't try to push down quals
+		 * to non-SELECT queries. There can be DML operation with RETURNING
```

Comment:
Acronym "DML" in some contexts also incorporates SELECT statements. Therefore, it's better to specify operation as "modifying DML".

Reply:
I think Postgres [docs](https://www.postgresql.org/docs/9.4/dml.html) are pretty clean about what data manipulation operations are. Thanks.